src/main/java/com/commercetools/sync/commons/utils/CommonTypeUpdateActionUtils.java

```diff
@@ -29,13 +31,57 @@
-            @Nullable final S oldObject,
-            @Nullable final S newObject,
-            @Nonnull final Supplier<U> updateActionSupplier) {
+        @Nullable final S oldObject,
+        @Nullable final S newObject,
+        @Nonnull final Supplier<U> updateActionSupplier) {
 
         return !Objects.equals(oldObject, newObject)
-                ? Optional.ofNullable(updateActionSupplier.get())
-                : Optional.empty();
+            ? Optional.ofNullable(updateActionSupplier.get())
+            : Optional.empty();
+    }
+
+    /**
+     * Compares two objects that are of type {@link ResourceIdentifier} (or a type that extends it) and returns a
+     * supplied {@link UpdateAction} as a result in an {@link Optional}. If both the {@link Object}s have the same
+     * values, then no update action is needed and hence an empty {@link Optional} is returned.
+     *
+     * @param oldResourceIdentifier the old resource identifier
+     * @param newResourceIdentifier the new resource identifier
+     * @param updateActionSupplier  the supplier that returns the update action to return in the optional
+     * @param <T>                   the type of the {@link UpdateAction}
+     * @param <S>                   the type of the old resource identifier
+     * @param <U>                   the type of the new resource identifier
+     * @param <V>                   concrete {@link UpdateAction} implementation type
+     * @return A filled optional with the update action or an empty optional if the object values are identical
+     */
+    @Nonnull
+    public static <T, S extends ResourceIdentifier, U extends ResourceIdentifier, V extends UpdateAction<T>> Optional<V>
+        buildUpdateActionForReferences(
+            @Nullable final S oldResourceIdentifier,
+            @Nullable final U newResourceIdentifier,
+            @Nonnull final Supplier<V> updateActionSupplier) {
+
+        return !areResourceIdentifiersEqual(oldResourceIdentifier, newResourceIdentifier)
+            ? Optional.ofNullable(updateActionSupplier.get())
+            : Optional.empty();
+    }
+
+    /**
+     * Compares the ids of two objects that are of type {@link ResourceIdentifier} (or a type that extends it).
+     *
+     * @param oldResourceIdentifier the old resource identifier
+     * @param newResourceIdentifier the new resource identifier
+     * @param <T>                   the type of the old resource identifier
+     * @param <S>                   the type of the new resource identifier
+     * @return true or false depending if the resource identifiers have the same id.
+     */
+    public static <T extends ResourceIdentifier, S extends ResourceIdentifier> boolean areResourceIdentifiersEqual(
+        @Nullable final T oldResourceIdentifier, @Nullable final S newResourceIdentifier) {
+
+        final String oldId = ofNullable(oldResourceIdentifier).map(ResourceIdentifier::getId).orElse(null);
```

Comment:
what happens if ResourceIdentifiers are key-based ? 
https://docs.commercetools.com/http-api-types.html#resourceidentifier-by-key

Reply:
This is not how the sync java should be used (as of now). Check the documentations:

https://commercetools.github.io/commercetools-sync-java/doc/usage/PRODUCT_SYNC/#sync-list-of-product-drafts

Reply:
hmm, I did not realize that before 👍